Fix volume reconstruction and add e2e tests

mkimuram opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
This PR fixes a bug that volume reconstruction does not work properly for csi, by failure in handling filesystem volume and block volume with a proper logic. To get the path to check for existence of volume, `Mounter` 's `GetPath()` should be used for file system and `BlockVolumeMapper`'s `GetPodDeviceMapPath()` should be used for block volume. In addition, `Mounter` should be created only for file system volume and `BlockVolumeMapper` should be created only for block volume, because csi plugin creates a new directory + json on `Mounter` and `BlockVolumeMapper` creation, leading to even more orphaned files/directories.
(This PR was originally intended to add e2e tests for block volume as discussed in https://github.com/kubernetes/kubernetes/issues/74545 , and found the bug and fix it.)

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:
/sig storage
cc @bswartz @wongma7 @msau42

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

mkimuram committed:
Move disruptive tests to testsuites and add ones for block volume

mkimuram commented:
@msau42 @wongma7 

Some of the fixes and additions to the existing test framework are borrowed from https://github.com/kubernetes/kubernetes/pull/74693. Also, changes that are discussed in https://github.com/kubernetes/kubernetes/pull/65248 are also included.

msau42 commented:
/ok-to-test

msau42 commented:
BTW, CSI Reconstruction is flaky in general: https://github.com/kubernetes/kubernetes/issues/72500

mkimuram commented:
/retest

mkimuram commented:
@msau42 

Is it possible to manually arrange tests in this PR, because they aren't tested by regular CI, due to the fact that these tests have ```[Disruptive]``` tag?

Note that below three tests are added for both filesystem and block volume in this PR.

```
"Should test that pv written before kubelet restart is readable after restart."
"Should test that pv used in a pod that is deleted while the kubelet is down cleans up when the kubelet returns."
"Should test that pv used in a pod that is force deleted while the kubelet is down cleans up when the kubelet returns."
```

For filesystem, all tests passed in my environment. (As they almost just call existing tests from framework.)
However, for block, only first test passes, and rest of the tests fail. (Maybe due to the issue discussed https://github.com/kubernetes/kubernetes/issues/74914)

Can this PR go ahead by disabling these failing tests for another PR that fix #74914 to utilize this test?
Or should we need to include fix to this PR?

msau42 commented:
/test pull-kubernetes-e2e-gce-csi-serial

Is an optional job that runs serial CSI tests

mkimuram commented:
@msau42 

> BTW, CSI Reconstruction is flaky in general: #72500

I found that the path is eventually deleted if manually checked after the test failure. Therefore, we might at least reduce the flake by adding rather longer wait before [here](https://sourcegraph.com/github.com/kubernetes/kubernetes/-/blob/test/e2e/storage/utils/utils.go#L251).

(In my environment, test with iscsi block volume fails more often in non-force delete case than in force delete case, so we might just need ~30sec wait in general and even more for CSI as a workaround?)

msau42 commented:
In the graceful delete case, kubelet should wait until volume manager finishes unmounting the volume before deleting the Pod object.  So we shouldn't need to add a timeout for the graceful deletion case.

mkimuram commented:
/test pull-kubernetes-e2e-gce-csi-serial

mkimuram commented:
/test pull-kubernetes-e2e-gce-csi-serial

msau42 commented:
/assign @bswartz

mkimuram commented:
@bswartz  @msau42 

Rebased and resolved conflict. PTAL

bswartz commented:
The change looks fine to me (I don't see any problem from reading the diff) but I'd like to actually run this change with one of my backends. What do I need to do to invoke the test suite and ensure that this new code is run?

mkimuram commented:
@bswartz 

> What do I need to do to invoke the test suite 

I haven't test with https://github.com/kubernetes/kubernetes/tree/master/test/e2e/storage/external by myself, yet. However, this should allow us to run storage e2e tests with your backend.

>and ensure that this new code is run?

And specifying `-focus` option makes specific tests to run.

mkimuram commented:
@bswartz

To make this test suite executable from external, we need one more line in the last commit.
Please try with the latest one.

Note that I was able to run this test suite with [modified version of csi-driver-nfs](https://github.com/kubernetes-csi/csi-driver-nfs/pull/13) on k8s cluster deployed with kubeadm, by below commands:

<details>

```
(After deployed modified version of csi-driver-nfs on k8s cluster.)
# cp go/src/github.com/kubernetes-csi/csi-driver-nfs/examples/kubernetes-simplenfs/storageclass.yaml /tmp
# cat /tmp/nfs-testdriver.yaml 
ShortName: mytest
StorageClass:
  FromName: false
  FromFile: /tmp/storageclass.yaml
DriverInfo:
  Name: csi-nfsplugin
  Capabilities:
    persistence: true
    exec: true
    dataSource: true
    multipods: true
    RWX: true
# cd go/src/k8s.io/kubernetes
# make all WHAT=test/e2e/e2e.test
# make ginkgo
# cd _output/local/bin/linux/amd64
# ./ginkgo -focus='External.Storage.*csi-nfsplugin.*Dynamic.*Should test that pv.*' e2e.test -- --kubeconfig=/root/.kube/config --provider=local --num-nodes=2 -storage.testdriver=/tmp/nfs-testdriver.yaml
```
</details>

(Nfs doesn't support block, so above doesn't allow me to test csi block, though.)

oomichi commented on test/e2e/framework/util.go:
```diff
@@ -4134,6 +4134,36 @@ func GetHostExternalAddress(client clientset.Interface, p *v1.Pod) (externalAddr
+	}
+
+	// If not found, return error
+	return "", fmt.Errorf("No external address for pod %v on node %v",
```
This function can return not only external one but also internal one.
So `external` should be removed from the error message.

oomichi commented on test/e2e/storage/testsuites/disruptive.go:
```diff
@@ -0,0 +1,156 @@
+				var err error
+				By("Creating a pod with pvc")
+				l.pod, err = framework.CreateSecPodWithNodeSelection(l.cs, l.ns.Name, []*v1.PersistentVolumeClaim{l.resource.pvc}, false, "", false, false, framework.SELinuxLabel, nil, framework.NodeSelection{Name: l.config.ClientNodeName}, framework.PodStartTimeout)
+				Expect(err).NotTo(HaveOccurred(), "While creating pods for kubelet restart test")
```
ditto

oomichi commented:
/cc @oomichi

mkimuram commented:
@oomichi 

Thank you for your review. I've fixed as suggested. PTAL

mkimuram commented:
The `pull-kubernetes-integration` test results for `csi-hostpath` show that there remains `"kubernetes.io~csi"` directory in `"/var/lib/kubelet/pods/{podUid}/volumeDevices/"` for `csi-hostpath` driver.
(And `csi-gce-driver` and other in-tree dirvers seem to somehow delete it, properly.)

I'm looking into it.

mkimuram commented:
@msau42 

I found that in-tree iscsi driver also hit the same issue for `force=false` case.

[UnmapDevice](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/volumepathhandler/volume_path_handler.go#L128) called from [unmapVolumeFunc](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/operationexecutor/operation_generator.go#L1202) will just remove the symbolic link, so it won't guarantee the parent path to be removed. (BTW, I manually confirmed that the path is eventually deleted. It seems to be handled in another place, and it causes a timing issue whether the code works before it is checked or not.) So, I think that we should check its child directory, by checking `"/var/lib/kubelet/pods/%s/volumeDevices/*/"`. 

By "Fix block volume reconstruction test to check podDevicePath instead of its parent" commit. Could you review it?
(It would be better to replace "*" to actual plugin directory name, in the future, but currently we have no information on plugin in the codes around there, so it is difficult to improve it right now.)

Note that it is at least fixed for in-tree iscsi driver like below:

<details>

```

[Before]
----
STEP: Expecting the symlinks from PodDeviceMapPath to be found.
Jun 13 16:04:30.471: INFO: ssh root@192.168.122.12:22: command:   sudo ls /var/lib/kubelet/pods/e49dd867-8df4-11e9-a4ef-525400b83be3/volumeDevices/ | grep '.'
Jun 13 16:04:30.471: INFO: ssh root@192.168.122.12:22: stdout:    "kubernetes.io~iscsi\n"
Jun 13 16:04:30.471: INFO: ssh root@192.168.122.12:22: stderr:    ""
Jun 13 16:04:30.471: INFO: ssh root@192.168.122.12:22: exit code: 0

<snip>

STEP: Expecting the symlink from PodDeviceMapPath not to be found.
Jun 13 16:05:26.896: INFO: ssh root@192.168.122.12:22: command:   sudo ls /var/lib/kubelet/pods/e49dd867-8df4-11e9-a4ef-525400b83be3/volumeDevices/ | grep '.'
Jun 13 16:05:26.896: INFO: ssh root@192.168.122.12:22: stdout:    "kubernetes.io~iscsi\n"
Jun 13 16:05:26.896: INFO: ssh root@192.168.122.12:22: stderr:    ""
Jun 13 16:05:26.896: INFO: ssh root@192.168.122.12:22: exit code: 0
----


[After]
----
STEP: Expecting the symlinks from PodDeviceMapPath to be found.
Jun 13 16:35:58.018: INFO: ssh root@192.168.122.12:22: command:   sudo ls /var/lib/kubelet/pods/4c0dfa21-8df9-11e9-a4ef-525400b83be3/volumeDevices/*/ | grep '.'
Jun 13 16:35:58.018: INFO: ssh root@192.168.122.12:22: stdout:    "iscsi-8rsj7\n"
Jun 13 16:35:58.018: INFO: ssh root@192.168.122.12:22: stderr:    ""
Jun 13 16:35:58.018: INFO: ssh root@192.168.122.12:22: exit code: 0

<snip>

STEP: Expecting the symlink from PodDeviceMapPath not to be found.
Jun 13 16:37:08.096: INFO: ssh root@192.168.122.12:22: command:   sudo ls /var/lib/kubelet/pods/4c0dfa21-8df9-11e9-a4ef-525400b83be3/volumeDevices/*/ | grep '.'
Jun 13 16:37:08.096: INFO: ssh root@192.168.122.12:22: stdout:    ""
Jun 13 16:37:08.096: INFO: ssh root@192.168.122.12:22: stderr:    "ls: cannot access '/var/lib/kubelet/pods/4c0dfa21-8df9-11e9-a4ef-525400b83be3/volumeDevices/*/': No such file or directory\n"
Jun 13 16:37:08.096: INFO: ssh root@192.168.122.12:22: exit code: 1
----
```

jsafrane commented:
/retest
with current csi-hostpath driver

jsafrane commented:
I'll check what's wrong and perhaps fix it.

jsafrane commented on test/e2e/storage/utils/utils.go:
```diff
@@ -281,6 +303,76 @@ func TestVolumeUnmountsFromForceDeletedPod(c clientset.Interface, f *framework.F
+	// Creating command to check whether path exists
+	command := fmt.Sprintf("ls /var/lib/kubelet/pods/%s/volumeDevices/*/ | grep '.'", clientPod.UID)
+	if isSudoPresent(nodeIP, framework.TestContext.Provider) {
+		command = fmt.Sprintf("sudo %s", command)
```
This needs to be `"sudo sh -c \"%s\""`. We need `volumeDevices/*/` to be expanded inside sudo shell, because the directory is accessible only to root.

jsafrane reviewed (commented):
I am trying to resurrect the PR and debug why it does not work in e2e. This is only the first finding, the tests still do not pass.

mkimuram commented:
@jsafrane 

> This is only the first finding, the tests still do not pass.

This seems not a test code bug, but an actual bug for csi plugin's reconstruction. 

In block volume case, it seems that csi plugin's [`NewMounter`](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_plugin.go#L440) is creating vol_data.json in [`reconstructVolume`]( https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/reconciler/reconciler.go#L496), by using the information created by [ReconstructVolumeOperation](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/reconciler/reconciler.go#L473).
And this causes (failure of) reconstruction for filesystem volume, even when it is for block volume.

So, we might need to avoid calling `NewMounter` in `reconstructVolume` for block volume case, or changing csi_plugin's `NewMounter` to take this path into account.

jsafrane committed:
Don't create mounter when reconstructing block volume

CSI mounter will create a new directory + json for a filesystem volume,
leading to even more orphaned files/directories.

jsafrane committed:
Add in-line filesystem volumes to MakeSecPod

jsafrane commented:
> So, we might need to avoid calling NewMounter in reconstructVolume for block volume case, or changing csi_plugin's NewMounter to take this path into account.

I created https://github.com/mkimuram/kubernetes/pull/1 (PR against your PR :-) ) that fixes the reconstruction. I'd like to have the tests and all fixes in one PR (or at least the fixes merged first before we merge broken test).

Most of the tests now pass, I am still chasing one random failure (see the PR).

mkimuram commented:
@jsafrane 

Thank you for your quick work!
It looks good to me at a glance. So, let me merge it to my PR and try CI testing.

I will also investigate the rest of issue that you shared.

jsafrane commented:
The tests are still failing, now because reconstruction of CSI filesystem volumes doesn't really work. See #79980 for details.

jsafrane commented on test/e2e/storage/testsuites/disruptive.go:
```diff
@@ -0,0 +1,161 @@
+		},
+		{
+			testItStmt:   "Should test that pv used in a pod that is deleted while the kubelet is down cleans up when the kubelet returns.",
+			runTestFile:  utils.TestVolumeUnmountsFromDeletedPod,
```
`runTestFile` here can't succeed because of #79980. Can you please comment it out with a proper TODO commit, so we can merge it as it is? I think that fixing the bug will take some time.

jsafrane commented:
@mkimuram, I see that my PR caused a merge commit, can you please rework the commits? I think it's ready for proper review and perhaps merge.
@msau42, PTAL

mkimuram commented:
@jsafrane @msau42 

I've squashed the commits and commented out the failing test with TODO comment.
PTAL

msau42 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -536,6 +514,31 @@ func (rc *reconciler) reconstructVolume(volume podVolume) (*reconstructedVolume,
 					newMapperErr)
 			}
 		}
+	} else {
```
Is having nil mounter safe to return in `reconstructedVolume`?

jsafrane replied:
That's goal of this PR. CSI's `NewMounter` would create directory and a json file for a **block** volume, which we don't want. `reconstructedVolume` should have either `mounter` **or** `mapper`, same as regular volumes in DSW.

jingxu97 replied:
just to make sure, for block volume, mounter is not needed at all.

mkimuram replied:
@jingxu97

Thank you for your comment.

> just to make sure, for block volume, mounter is not needed at all.

As I checked operation_executor/operation_generator code paths for block volume, mounter is not directly used in these paths, except [`GenerateExpandInUseVolumeFunc`](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/operationexecutor/operation_generator.go#L1633) .
And I guess that this path won’t be reached for block volume, because `FSResizeRequired` won’t be set due to [this check](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go#L374).

Is my understanding right? @gnufied @mlmhl

Also, could you double check if we should check any other paths deeper? @jsafrane

```
AttachVolume  --> GenerateAttachVolumeFunc --> <Attacher>
DetachVolume --> GenerateDetachVolumeFunc --> <Detacher>
MountVolume --> GenerateMapVolumeFunc --> <BlockVolumeMapper>
UnmountVolume --> GenerateUnmapVolumeFunc --> <BlockVolumeUnmapper>
UnmountDevice --> GenerateUnmapDeviceFunc --> <BlockVolumeUnmapper>

ExpandInUseVolume --> GenerateExpandInUseVolumeFunc --> <Attacher, Mounter>

VerifyVolumesAreAttached --> GenerateBulkVolumeVerifyFunc --> <Attacher>
VerifyVolumesAreAttachedPerNode --> GenerateVolumesAreAttachedFunc --> <Attacher>
VerifyControllerAttachedVolume --> GenerateVerifyControllerAttachedVolumeFunc --> <None>

ReconstructVolumeOperation --> <None>
CheckVolumeExistenceOperation --> <None>
```

jsafrane replied:
@mkimuram, thanks for your analysis! It confirms my finding that mounter is not really needed for block volumes.

jingxu97 replied:
I was checking here, and thought there might be issue. But since it only loops over container.VolumeMounts which does not contain blockvolumes, it should be fine then.
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_pods.go#L146

msau42 commented:
/assign @jingxu97 
to review reconstruction changes

mkimuram commented:
Rebased.

@jingxu97

Could you review reconstruction changes for this PR? (Reconstruction part of code is the second commit "Don't create mounter when reconstructing block volume". And related discussion starts from https://github.com/kubernetes/kubernetes/pull/75071#issuecomment-509457591 )

mkimuram commented:
Rebased again to resolve merge conflict with #78000 .
The last commit is to apply the same change in "Add in-line filesystem volumes to MakeSecPod" commit to testsuites/volume_expand.go, which is newly added in the conflicted PR.

PTAL

@jsafrane @msau42 @jingxu97

jsafrane commented:
I probably should not lgtm PRs that have my commits, but it's good to be merged.
/approve

msau42 commented:
/approve

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -493,32 +493,10 @@ func (rc *reconciler) reconstructVolume(volume podVolume) (*reconstructedVolume,
-	var volumeMapper volumepkg.BlockVolumeMapper
 	if utilfeature.DefaultFeatureGate.Enabled(features.BlockVolume) && volume.volumeMode == v1.PersistentVolumeBlock {
 		var newMapperErr error
 		if mapperPlugin != nil {
```
It is not your change, but I don't see the possibility of mapperPlugin is nil?

mkimuram replied:
@jingxu97 

Thank you for your comment.

You're right that this path won't be reached, beause [`ReconstructVolumeOperation` returns error when `mapperPlugin == nil`](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/operationexecutor/operation_executor.go#L901) just after [`FindMapperPluginByName` returns `nil` for `mapperPlugin`](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/plugins.go#L1022) when plugin doesn't support block volume.

However, it doesn't seem to be so obvious, because it depends on other function's implementation. 
In this case, do you mean to delete this check here or any other ideas for improve this?

jingxu97 replied:
It just seems strange to check this quite late here.

mkimuram replied:
@jingxu97 @oomichi 

I've moved nil check for `mapperPlugin` earlier in `reconstructVolume` in "Move nil check for mapperPlugin earlier in reconstructVolume" commit. PTAL

(I will merge this change with the previous commit if this looks good.)

jsafrane commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -536,6 +514,31 @@ func (rc *reconciler) reconstructVolume(volume podVolume) (*reconstructedVolume,
+	}
+
+	// Check existence of mount point for filesystem volume or symbolic link for block volume
+	isExist, checkErr := rc.operationExecutor.CheckVolumeExistenceOperation(volumeSpec, volume.volumePath, volumeSpec.Name(), rc.mounter, uniqueVolumeName, volume.podName, pod.UID, attachablePlugin)
```
`volume.volumePath` is used incorrectly here. Before this PR, `volumeMounter.GetPath()` was used. This breaks CSI reconstruction, as `CheckVolumeExistenceOperation` checks the pod directory [1] and not the directory where the volume should be mounted [2]

1: `/var/lib/kubelet/pods/4ec8e3aa-f83e-4d49-96ea-3859f09f7dd7/volumes/kubernetes.io~csi/pvc-873e54a8-a2c1-4872-8a3c-b68b71a81e42`
2: `/var/lib/kubelet/pods/4ec8e3aa-f83e-4d49-96ea-3859f09f7dd7/volumes/kubernetes.io~csi/pvc-873e54a8-a2c1-4872-8a3c-b68b71a81e42/mount`

jsafrane commented:
@mkimuram, I fixed CSI volume reconstruction in the last commit of this PR: https://github.com/kubernetes/kubernetes/pull/80743 (running tests there).

Can you please include it in this PR? I am sorry I broke it in the first place.

mkimuram commented:
@jsafrane

Oh, I should have read through the original code, too. Sorry for that.
Previous code seems to revert #74653 work (@jingxu97 was right) and now it is taken into account.
Also, checking `GetPodDeviceMapPath` for block volume should be the right place to check.

I've re-organized the code to move the fix into the right commits.
(Deleted the test disable commit and merged the fix commit into one.)

PTAL @jsafrane @msau42 @jingxu97

mkimuram commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -493,32 +493,10 @@ func (rc *reconciler) reconstructVolume(volume podVolume) (*reconstructedVolume,
-		return nil, fmt.Errorf("Volume: %q is not mounted", uniqueVolumeName)
-	}
+	var volumeMapper volumepkg.BlockVolumeMapper
+	var volumeMounter volumepkg.Mounter
```
@jingxu97 

Thank you for your review.

I've changed the PR's title and comments and added comments in `MountedVolume` definition.
(New code change is merged to "Don't create mounter when reconstructing block volume" commit.)

PTAL

oomichi commented on test/e2e/storage/generic_persistent_volume-disruptive.go:
```diff
@@ -102,7 +102,9 @@ func createPodPVCFromSC(f *framework.Framework, c clientset.Interface, ns string
-	pod, err := framework.CreateNginxPod(c, ns, nil, pvcClaims)
+	pod, err := framework.CreateSecPod(c, ns, pvcClaims, nil,
+		false, "", false, false, framework.SELinuxLabel,
+		nil, framework.PodStartTimeout)
```
Why do we need to change here?
I cannot find the reason from the PR message.

mkimuram replied:
`framework.CreateNginxPod` can't handle block volume and inline volume, but `framework.CreateSecPod` can.
Current tests in generic_persistent_volume-disruptive.go doesn't actually test them, but it would be good to use the similar logic to testsuites/disruptive.go to prepare for future imporvement and maintenance.

oomichi commented on test/e2e/storage/utils/utils.go:
```diff
@@ -289,6 +311,76 @@ func TestVolumeUnmountsFromForceDeletedPod(c clientset.Interface, f *framework.F
+	result, err := e2essh.SSH(command, nodeIP, framework.TestContext.Provider)
+	e2essh.LogResult(result)
+	framework.ExpectNoError(err, "Encountered SSH error.")
+	gomega.Expect(result.Code).To(gomega.BeZero(), fmt.Sprintf("Expected grep exit code of 0, got %d", result.Code))
```
We can replace here with `framework.ExpectEqual(result.Code, 0, fmt.Sprintf("Expected grep exit code of 0, got %d", result.Code))`

mkimuram replied:
Fixed as suggested.

oomichi commented on test/e2e/storage/utils/utils.go:
```diff
@@ -289,6 +311,76 @@ func TestVolumeUnmountsFromForceDeletedPod(c clientset.Interface, f *framework.F
+	ginkgo.By("Starting the kubelet and waiting for pod to delete.")
+	KubeletCommand(KStart, c, clientPod)
+	err = f.WaitForPodNotFound(clientPod.Name, framework.PodDeleteTimeout)
+	if err != nil {
```
We can remove `if err != nil {` because of `framework.ExpectNoError(err, "Expected pod to be not found.")`

mkimuram replied:
Fixed as suggested.

mkimuram committed:
Refactor e2e tests

mkimuram committed:
Move nil check for mapperPlugin earlier in reconstructVolume

mkimuram reviewed (commented):
@oomichi 

Thank you for your comment. I've fixed as suggested. PTAL

jsafrane commented:
/retest

jingxu97 commented:
thanks for the work! /lgtm

jsafrane commented:
Just reformating Jing's lgtm
/lgtm
